python_modules/libraries/dagster-components/dagster_components/core/component.py

```diff
@@ -204,12 +207,55 @@ def get_component_types_in_module(
 T = TypeVar("T")
 
 
+@dataclass
+class MultiComponentsLoadContext:
+    """Context for loading an entire component hierarchy in a code location.
+    Stores resources and a cache to ensure we don't load the same component multiple times.
+    """
+
+    resources: Mapping[str, object]
+
+    def build_defs_from_component_path(self, path: Path) -> Definitions:
+        """Build a definitions object from a folder within the components hierarchy."""
+        return self._build_defs_from_component_path_inner(path.absolute().resolve())
+
+    @cached_method
+    def _build_defs_from_component_path_inner(self, path: Path) -> Definitions:
```

Comment:
And this cache should be shared with the mainline loading pathways as well. We should have a canonical cache that maps module_name to definitions object. Or, even more spicy, we could monkey patch the python module itself to have the defs object

Reply:
It is shared with the main loading pathways, at least `build_component_defs`, which constructs the outer cache-storing context before loading individual components.